modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -127,6 +133,14 @@ public void clusterChanged(ClusterChangedEvent event) {
                 cancelJob();
                 // clear the deduplicator on master failover so we could re-send the requests in case we're re-elected
                 transportActionsDeduplicator.clear();
+                errorStore.clearStore();
+            }
+        }
+        if (event.localNodeMaster()) {
+            // only execute if we're the master
+            List<Index> indicesDeleted = event.indicesDeleted();
+            for (Index deleted : indicesDeleted) {
+                errorStore.clearRecordedError(deleted.getName());
             }
```

Comment:
I think we could do this differently, since we don't necessarily need to "clear" this on every cluster state update, it would be fine for it to go away on the next DLM interval.

For example, we could change this method:

```java
    private void clearErrorStoreForUnmanagedIndices(DataStream dataStream) {
        Metadata metadata = clusterService.state().metadata();
        for (Index index : dataStream.getIndices()) {
            if (dataStream.isIndexManagedByDLM(index, metadata::index) == false) {
                errorStore.clearRecordedError(index.getName());
            }
        }
    }
```

To one that does something like:

```java
    private void clearErrorStoreForUnmanagedIndices(DataStream dataStream) {
        Metadata metadata = clusterService.state().metadata();
        for (String index : errorStore.getAllIndices()) {
            if (metadata.index(index) == null) {
                errorStore.clearRecordedError(index);
            } else if (dataStream.isIndexManagedByDLM(index, metadata::index) == false) {
                errorStore.clearRecordedError(index.getName());
            }
        }
    }
```

And then it would do garbage collection for anything that was deleted manually or otherwise, and since it's based on a listing from the errorStore, there's no possibility that it gets out of sync.

Reply:
++ great suggestion, thanks Lee
Coming up in a separate PR